src/libtorch.cc

```diff
@@ -1265,6 +1284,96 @@ ModelInstanceState::Execute(
   }
 }
 
+TRITONSERVER_Error*
+ModelInstanceState::GetNamingConvention(
+    NamingConvention* naming_convention,
+    const std::set<std::string>& allowed_ios)
+{
+  // Rules for (non-Dictionary) input tensor names:
+  // 1. Must be in 'allowed_inputs' (arguments in the forward function)
+  // 2. Must follow the naming convention i.e. <name>__<index>
+  // 3. If neither of the above conditions are satisfied, enforce strict
+  // ordering of model inputs.
+  //
+  // Rules for output tensor names:
+  // 1. Must follow the naming convention i.e. <name>__<index>
+  // 2. If not, we enforce strict ordering of model outputs.
+  std::string deliminator = "__";
+  size_t io_index = 0;
+  std::string io_kind = "input";
+  *naming_convention = NamingConvention::FORWARD_ARGUEMENT;
+  if (allowed_ios.size() == 0) {
+    // symbolizes output
+    io_kind = "output";
+    *naming_convention = NamingConvention::NAMED_INDEX;
+  }
+
+  triton::common::TritonJson::Value ios;
+  RETURN_IF_ERROR(
+      model_state_->ModelConfig().MemberAsArray(io_kind.c_str(), &ios));
+
+  if (io_kind == "input") {
+    for (size_t i = 0; i < ios.ArraySize(); i++) {
+      triton::common::TritonJson::Value io;
+      RETURN_IF_ERROR(ios.IndexAsObject(i, &io));
+
+      // Validate name
+      std::string io_name;
+      RETURN_IF_ERROR(io.MemberAsString("name", &io_name));
+      auto itr = allowed_ios.find(io_name);
+      if (itr == allowed_ios.end()) {
+        *naming_convention = NamingConvention::NAMED_INDEX;
+        break;
+      }
+    }
+  }
+
+  // If not, check if inputs follow INDEX
+  if (*naming_convention == NamingConvention::NAMED_INDEX) {
+    for (size_t i = 0; i < ios.ArraySize(); i++) {
+      triton::common::TritonJson::Value io;
+      RETURN_IF_ERROR(ios.IndexAsObject(i, &io));
+
+      // Validate name
+      std::string io_name;
+      RETURN_IF_ERROR(io.MemberAsString("name", &io_name));
+      try {
+        int start_pos = io_name.find(deliminator);
+        if (start_pos == -1) {
+          throw std::invalid_argument(
+              "input/output must follow naming convention");
```

Comment:
NIT: try-catching in c++ is pretty expensive even if this is as straight forward as it looks. I tried to make a [quick example](https://godbolt.org/z/hfTGceErK) of what the output of the assembler looks like. You can see that gcc has some more bloat with the exception handling as compared to the standard if-statement error handling. Turning on -O3 didn't really improve things either. If we expect multiple types of exceptions to be thrown in this block then we can keep it but otherwise I would suggest using the standard if-statements to handle this particular case.

Reply:
I think is a great idea. I have re-worked to avoid the catching of exceptions for std::aoti